descwl_shear_testing/mbobs_extractor.py

```diff
@@ -0,0 +1,422 @@
+        rec: an object record
+            TODO I don't actually know what class this is
+
+        returns
+        --------
+        obs: ngmix.Observation
+            The Observation, including
+        """
+
+        im = imobj_sub.image.array
+        wt = self._extract_weight(imobj_sub)
+        maskobj = imobj_sub.mask
+        bmask = maskobj.array
+        jacob = self._extract_jacobian(imobj_sub, rec)
+
+        # cen = rec.getCentroid()
+        orig_cen = imobj_sub.getWcs().skyToPixel(rec.getCoord())
+
+        if psf_im is None:
+            psf_im = self._extract_psf_image(imobj_sub, orig_cen)
+
+        # fake the psf pixel noise
+        psf_err = psf_im.max()*0.0001
+        psf_wt = psf_im*0 + 1.0/psf_err**2
+
+        # use canonical center for the psf
+        psf_cen = (np.array(psf_im.shape)-1.0)/2.0
+        psf_jacob = jacob.copy()
+        psf_jacob.set_cen(row=psf_cen[0], col=psf_cen[1])
+
+        # we will have need of the bit names which we can only
+        # get from the mask object
+        # this is sort of monkey patching, but I'm not sure of
+        # a better solution
+        meta = {'maskobj': maskobj}
+
+        psf_obs = ngmix.Observation(
+            psf_im,
+            weight=psf_wt,
+            jacobian=psf_jacob,
+        )
+        obs = ngmix.Observation(
+            im,
+            weight=wt,
+            bmask=bmask,
+            jacobian=jacob,
+            psf=psf_obs,
+            meta=meta,
+        )
+
+        return obs
+
+    def _extract_psf_image(self, stamp, orig_pos):
+        """
+        get the psf associated with this stamp
+
+        coadded psfs are generally not square, so we will
+        trim it to be square and preserve the center to
+        be at the new canonical center
+        """
+        try:
+            psfobj = stamp.getPsf()
+            psfim = psfobj.computeKernelImage(orig_pos).array
+        except InvalidParameterError:
+            raise MBObsMissingDataError("could not reconstruct PSF")
+
+        psfim = np.array(psfim, dtype='f4', copy=False)
+
+        psfim = util.trim_odd_image(psfim)
+        return psfim
+
+    def _extract_jacobian(self, imobj, rec):
+        """
+        extract an ngmix.Jacobian from the image object
+        and object record
+
+        imobj: an image object
+            TODO I don't actually know what class this is
+        rec: an object record
+            TODO I don't actually know what class this is
+
+        returns
+        --------
+        Jacobian: ngmix.Jacobian
+            The local jacobian
+        """
+
+        xy0 = imobj.getXY0()
+
+        orig_cen = imobj.getWcs().skyToPixel(rec.getCoord())
+        cen = orig_cen - geom.Extent2D(xy0)
+        row = cen.getY()
+        col = cen.getX()
+
+        wcs = imobj.getWcs().linearizePixelToSky(
+            orig_cen,
+            geom.arcseconds,
+        )
+        jmatrix = wcs.getLinear().getMatrix()
+
+        jacob = ngmix.Jacobian(
```

Comment:
I must have the convention wrong here and this is causing the negative responses